pennylane/measurements/sample.py

```diff
@@ -182,9 +182,33 @@ def __init__(self, obs=None, wires=None, eigvals=None, id=None):
 
         super().__init__(obs=obs, wires=wires, eigvals=eigvals, id=id)
 
-    @property
-    def return_type(self):
-        return Sample
+    return_type = Sample
+
+    @classmethod
+    def _abstract_eval(
+        cls,
+        n_wires: Optional[int] = None,
+        has_eigvals=False,
+        shots: Optional[int] = None,
+        num_device_wires: int = 0,
+    ):
+        if shots is None:
+            raise ValueError("finite shots are required to SampleMP")
+        dtype = float if n_wires is None or has_eigvals else int
+
+        if n_wires == 0:
+            dim = num_device_wires
+        elif has_eigvals or n_wires is None:
```

Comment:
Maybe introduce this flag for readability:

```suggestion
        sample_eigvals = n_wires is None or has_eigvals
        dtype = float if sample_eigvals else int

        if n_wires == 0:
            dim = num_device_wires
        elif sample_eigvals:
```